Presto not respecting SSL and always uses http instead of https

**Describe the bug**
Presto not respecting SSL setting and always uses `http` instead of `https`, which means it cannot connect to databases that only allows https.

**To Reproduce**
1. Admin > Databases > Add database > Presto and fill out information including SSL activated:
![image](https://user-images.githubusercontent.com/1447303/93924857-99851e00-fd15-11ea-856d-996f1a993d7f.png)
2. When saving, Metabase will try to connect to `http://presto.example.test:443` instead of over `https` - the log would output something like this:
`Caused by: clojure.lang.ExceptionInfo: clj-http: status 404 {:object {:request-time 5, :repeatable? false, :protocol-version {:name "HTTP", :major 1, :minor 1}, :streaming? true, :chunked? false, :reason-phrase "Not Found", :headers {"Content-Type" "text/plain; charset=utf-8", "X-Content-Type-Options" "nosniff", "Date" "Tue, 22 Sep 2020 20:21:25 GMT", "Content-Length" "19", "Connection" "close"}, :orig-content-encoding nil, :status 404, :length 19, :body "404 page not found\n", :trace-redirects ["http://presto.example.test:443/v1/statement"]}, :environment {req {:headers {"X-Presto-Source" "test", :as :json, :redirect-strategy :lax, :url "http://presto.example.test:443/v1/statement", :request-method :post}, p__23212 {:request-time 5, :repeatable? false, :protocol-version {:name "HTTP", :major 1, :minor 1}, :streaming? true, :chunked? false, :reason-phrase "Not Found", :headers {"Content-Type" "text/plain; charset=utf-8", "X-Content-Type-Options" "nosniff", "Date" "Tue, 22 Sep 2020 20:21:25 GMT", "Content-Length" "19", "Connection" "close"}, :orig-content-encoding nil, :status 404, :length 19, :body "404 page not found\n", :trace-redirects ["http://presto.example.test:443/v1/statement"]}, map__23213 {:request-time 5, :repeatable? false, :protocol-version {:name "HTTP", :major 1, :minor 1}, :streaming? true, :chunked? false, :reason-phrase "Not Found", :headers {"Content-Type" "text/plain; charset=utf-8", "X-Content-Type-Options" "nosniff", "Date" "Tue, 22 Sep 2020 20:21:25 GMT", "Content-Length" "19", "Connection" "close"}, :orig-content-encoding nil, :status 404, :length 19, :body "404 page not found\n", :trace-redirects ["http://presto.example.test:443/v1/statement"]}, resp {:request-time 5, :repeatable? false, :protocol-version {:name "HTTP", :major 1, :minor 1}, :streaming? true, :chunked? false, :reason-phrase "Not Found", :headers {"Content-Type" "text/plain; charset=utf-8", "X-Content-Type-Options" "nosniff", "Date" "Tue, 22 Sep 2020 20:21:25 GMT", "Content-Length" "19", "Connection" "close"}, :orig-content-encoding nil, :status 404, :length 19, :body "404 page not found\n", :trace-redirects ["http://presto.example.test:443/v1/statement"]}, status 404}}
`

**Information about your Metabase Installation:**
Metabase 1.36.6.1 and 0.36.6

**Additional context**
From the code, it should automatically select `https`, when SSL is activated:
https://github.com/metabase/metabase/blob/d48accbeadfebd3ef4bb0a128408a98cd0fdf94e/modules/drivers/presto/src/metabase/driver/presto.clj#L70-L73

--- a/src/metabase/api/database.clj
+++ b/src/metabase/api/database.clj
@@ -437,21 +437,26 @@
 (s/defn ^:private test-connection-details :- su/Map
   "Try a making a connection to database `engine` with `details`.
 
-  Tries twice: once with SSL, and a second time without if the first fails. If either attempt is successful, returns
+  If the `details` has SSL explicitly enabled, go with that and do not accept plaintext connections. If it is disabled,
+  try twice: once with SSL, and a second time without if the first fails. If either attempt is successful, returns
   the details used to successfully connect. Otherwise returns a map with the connection error message. (This map will
   also contain the key `:valid` = `false`, which you can use to distinguish an error from valid details.)"
   [engine :- DBEngineString, details :- su/Map]
-  (let [details (if (supports-ssl? (keyword engine))
-                  (assoc details :ssl true)
-                  details)]
+  (if (and (supports-ssl? (keyword engine))
+           (true? (:ssl details)))
+    (let [error (test-database-connection engine details)]
+      (or error details))
+    (let [details (if (supports-ssl? (keyword engine))
+                    (assoc details :ssl true)
+                    details)]
     ;; this loop tries connecting over ssl and non-ssl to establish a connection
     ;; if it succeeds it returns the `details` that worked, otherwise it returns an error
-    (loop [details details]
-      (let [error (test-database-connection engine details)]
-        (if (and error
-                 (true? (:ssl details)))
-          (recur (assoc details :ssl false))
-          (or error details))))))
+      (loop [details details]
+        (let [error (test-database-connection engine details)]
+          (if (and error
+                   (true? (:ssl details)))
+            (recur (assoc details :ssl false))
+            (or error details)))))))
 
 (def ^:private CronSchedulesMap
   "Schema with values for a DB's schedules that can be put directly into the DB."

--- a/test/metabase/api/database_test.clj
+++ b/test/metabase/api/database_test.clj
@@ -683,7 +683,34 @@
           (testing "via the API endpoint"
             (is (= {:valid false}
                    ((mt/user->client :crowberto) :post 200 "database/validate"
-                    {:details {:engine :h2, :details {:db "ABC"}}})))))))))
+                    {:details {:engine :h2, :details {:db "ABC"}}})))))))
+
+    (let [call-count (atom 0)
+          ssl-values (atom [])]
+      (with-redefs [database-api/test-database-connection (fn [_ details]
+                                                            (swap! call-count inc)
+                                                            (swap! ssl-values conj (:ssl details))
+                                                            {:valid true})]
+          (testing "with SSL enabled, do not allow non-SSL connections"
+            (#'database-api/test-connection-details "presto" {:ssl true})
+            (is (= 1 @call-count))
+            (is (= [true] @ssl-values)))
+
+          (reset! call-count 0)
+          (reset! ssl-values [])
+
+          (testing "with SSL disabled, try twice (once with, once without SSL)"
+            (#'database-api/test-connection-details "presto" {:ssl false})
+            (is (= 2 @call-count))
+            (is (= [true false] @ssl-values)))
+
+          (reset! call-count 0)
+          (reset! ssl-values [])
+
+          (testing "with SSL unspecified, try twice (once with, once without SSL)"
+            (#'database-api/test-connection-details "presto" {})
+            (is (= 2 @call-count))
+            (is (= [true false] @ssl-values)))))))
 
 
 ;;; +----------------------------------------------------------------------------------------------------------------+

